src/services/suggestionDiagnostics.ts

```diff
@@ -24,20 +24,30 @@ namespace ts {
                             diags.push(createDiagnosticForNode(isVariableDeclaration(node.parent) ? node.parent.name : node, Diagnostics.This_constructor_function_may_be_converted_to_a_class_declaration));
                         }
                     }
-                    
-                    if(node.modifiers && containsAsync(node.modifiers)){
+                    if (node.modifiers && isAsyncFunction(node)) {
                         break;
                     }
 
-                    const returnType = checker.getReturnTypeOfSignature(checker.getSignatureFromDeclaration(<FunctionDeclaration | FunctionExpression> node))
-                    if(!returnType || !returnType.symbol){
+                    const returnType = checker.getReturnTypeOfSignature(checker.getSignatureFromDeclaration(<FunctionDeclaration | FunctionExpression>node));
+                    if (!returnType || !returnType.symbol) {
                         break;
                     }
 
-                    if(isPromiseType(returnType)){
+                    // collect all the return statements
+                    // check that a property access expression exists in there and that it is a handler
+                    const retStmts: ReturnStatement[] = [];
+                    getReturnStmts(node, retStmts);

```

Comment:
i would recommend having all this logic in one function, and sharing it between the quick fix and the suggestion code. this way we can use them to report the suggestion, and check in the quick fix if the assumptions are still valid.

this comes up when VSCode, say call us to apply a fix after a code change, so it is possible that the previous code where we reported the suggestion has change.